### PR TITLE
User delegate at deeper

### PR DESCRIPTION
**Pull request by jerry-yu**

### CODE REVIEW

File: pallets/credit/src/lib.rs [hunk 1]
```diff
@@ -728,17 +837,64 @@ pub mod pallet {
...
+                old_data.reward_eras = DEFAULT_REWARD_ERAS;
+            }
 
-            UserCredit::<T>::insert(who, old_data.clone());
+            UserCredit::<T>::insert(who, old_data);
             Self::update_credit_history(who, expire_era);
             true
         }
+
+        fn do_add_credit(account_id: T::AccountId, credit_data: CreditData) -> DispatchResult {
```
Comment (xcaptain) [inline pallets/credit/src/lib.rs:854]: This method won't return `Err`, how about not return a `Result`

File: pallets/credit/src/lib.rs [hunk 2]
```diff
@@ -728,17 +837,64 @@ pub mod pallet {
...
     impl<T: Config> CreditInterface<T::AccountId, BalanceOf<T>> for Pallet<T> {
+        fn get_credit_balance() -> Vec<BalanceOf<T>> {
+            Self::credit_balances()
+        }
+
+        fn get_credit_gap(dst_lv: u8, cur_lv: u8) -> u64 {
+            CreditLevel::credit_level_gap(dst_lv.into(), cur_lv.into())
+        }
+
+        fn add_or_update_credit(account_id: T::AccountId, credit_score: u64) -> DispatchResult {
```
Comment (xcaptain) [inline pallets/credit/src/lib.rs:880]: No `Err` was returned, consider not return a `Result`

Comment (jerry-yu) [reply pallets/credit/src/lib.rs:890]: good

File: pallets/credit/src/lib.rs (hunk 1 above)
Comment (jerry-yu) [reply pallets/credit/src/lib.rs:864]: good

File: pallets/staking/src/lib.rs [hunk 3]
```diff
@@ -2079,6 +2066,89 @@ pub mod migrations {
 }
 
 impl<T: Config> pallet::Pallet<T> {
+    pub fn account_id() -> T::AccountId {
+        T::PalletId::get().into_account()
```
[the remark this answers is not in the document]
Comment (jerry-yu) [reply pallets/staking/src/lib.rs:2072]: it it safe  when you only tell somebody your public key

File: pallets/credit/src/lib.rs [hunk 4]
```diff
@@ -251,6 +334,35 @@ pub mod pallet {
...
 
+    #[pallet::storage]
+    #[pallet::getter(fn credit_balances)]
+    pub type CreditBalances<T: Config> =
+        StorageValue<_, Vec<BalanceOf<T>>, ValueQuery, CreditDefaultBalance<T>>;
+
+    #[pallet::type_value]
+    pub fn CreditDefaultBalance<T: Config>() -> Vec<BalanceOf<T>> {
+        vec![
+            UniqueSaturatedFrom::unique_saturated_from(1000 * DPR),
```
[the remark this answers is not in the document]
Comment (jerry-yu) [reply pallets/credit/src/lib.rs:345]: type not match

File: pallets/credit/src/lib.rs [hunk 5]
```diff
@@ -251,6 +334,35 @@ pub mod pallet {
...
+    }
+
+    #[pallet::type_value]
+    pub fn NewUserCampaignId() -> u16 {
+        4
+    }
+
+    #[pallet::storage]
+    #[pallet::getter(fn default_campaign_id)]
+    pub(crate) type DefaultCampaignId<T> = StorageValue<_, u16, ValueQuery, NewUserCampaignId>;
```
[the remark this answers is not in the document]
Comment (jerry-yu) [reply pallets/credit/src/lib.rs:364]: it is used

File: pallets/credit/src/lib.rs [hunk 6]
```diff
@@ -251,6 +334,35 @@ pub mod pallet {
     pub type NotSwitchAccounts<T: Config> =
         StorageMap<_, Blake2_128Concat, T::AccountId, bool, OptionQuery>;
 
+    #[pallet::storage]
+    #[pallet::getter(fn credit_balances)]
+    pub type CreditBalances<T: Config> =
```
[the remark this answers is not in the document]
Comment (jerry-yu) [reply pallets/credit/src/lib.rs:339]: If setting constant, it must doing upgrade when need changing

File: pallets/staking/src/lib.rs [hunk 7]
```diff
@@ -990,6 +999,48 @@ pub mod pallet {
...
+                ExistenceRequirement::KeepAlive,
+            )?;
+            DelegatorBalances::<T>::mutate(&who, |balance| {
+                *balance = balance.saturating_add(need_balance)
+            });
+            T::CreditInterface::add_or_update_credit(
+                who.clone(),
+                credit_score.unwrap_or(0) + score_gap,
+            );
+            Self::delegate_any(who)?;
```
[the remark this answers is not in the document]
Comment (jerry-yu) [reply pallets/staking/src/lib.rs:1040]: Delegating must doing after credit updating

File: pallets/staking/src/tests.rs [hunk 8]
```diff
@@ -2968,3 +2968,51 @@ fn elected_validators_should_rotate() {
...
 }
+
+#[test]
+fn staking_delegate() {
+    ExtBuilder::default()
+        .validator_pool(true)
+        .build_and_execute(|| {
+            assert_ok!(Credit::set_credit_balances(
+                Origin::root(),
+                vec![
```
[the remark this answers is not in the document]
Comment (jerry-yu) [reply pallets/staking/src/tests.rs:2979]: using constant ,you can't change in test env

**Review by xcaptain — APPROVED** [review]

**Pull request merged**